examples/slackbot/Notebook.ipynb

```diff
@@ -27,89 +27,93 @@
+    "    conversation_start = since_last_message > k.minutes(5)\n",
     "\n",
-    "# Create a new local session\n",
-    "s = k.session()\n",
+    "    k.make_record( \n",
+    "        # A list of all messages over the past 10 minutes (up to 100)\n",
+    "        recent_messages=messages\n",
+    "            .select(\"user\", \"type\", \"text\")\n",
+    "            .collect(window=since(conversation_start), max=100),\n",
     "\n",
-    "# Pull live history into Astra Streaming\n",
-    "# -> Airbyte to Pulsar\n",
-    "# -> https://docs.airbyte.com/integrations/sources/slack\n",
-    "# -> https://docs.airbyte.com/integrations/destinations/pulsar\n",
+    "        # How many messages have been reacted to in the conversation\n",
+    "        reaction_count=messages\n",
+    "            .when(messages.col(\"reactions\").is_not_null())\n",
+    "            .count(window=since(conversation_start)),\n",
+    "    )\n",
     "\n",
-    "# Load messages into Kaskada\n",
-    "# For training: message = k.read_parquet(..., entity=\"user_id\")\n",
-    "message = s.read_pulsar(..., entity=\"user_id\")\n",
+    "def examples(messages):\n",
+    "    k.make_record(\n",
+    "        prompt=prompt(messages).lag(1).lookup(messages.get(\"channel\")),\n",
     "\n",
-    "# Build prompt\n",
-    "messages_by_channel = message.with_key(message.channel)\n",
-    "prompt = k.make_record( \n",
-    "  # A list of all messages over the past 10 minutes (up to 100)\n",
-    "  recent_messages=messages_by_channel \n",
-    "    .select(\"user\", \"type\", \"text\")\n",
-    "    .last_duration(timedelta(minutes=10), max=100),\n",
-    "  # How many messages have been reacted to in the last 1 minutes\n",
-    "  engage_count=messages_by_channel\n",
-    "    .when(messages_by_channel.get(\"reactions\").is_not_null())\n",
-    "    .count(window=duration(timedelta(minutes=10))),\n",
-    "  # The timestamp of the last message\n",
-    "  time=time_of(messages_by_channel),\n",
-    ")\n",
+    "        # We'll train ChatGPT to generate the user ID who will engage next\n",
+    "        completion=completion.col(\"user_id\"),\n",
+    "    )\n",
     "\n",
-    "# We'll train ChatGPT to generate the user ID who will engage next\n",
-    "completion = \"\\n\\n###\\n\\n\" + message.get(\"user_id\")\n",
+    "def main():\n",
+    "    # Initialize Kaskada with a local execution context.\n",
+    "    k.init_session()\n",
     "\n",
-    "# Each example will predict who will engage next, given\n",
-    "# the last 10 minutes of conversation history\n",
-    "examples = k.make_record(\n",
-    "  prompt=prompt.lag(1).lookup(message.get(\"channel\")),\n",
-    "  completion=completion,\n",
-    ")\n",
-    "examples = examples.when(examples.get(\"prompt\").is_not_null())\n",
+    "    tl = examples(messages = k.ParquetSource(files=[\"./messages.parquet\"]))\n",
+    "    tl = tl.when(tl.col(\"prompt\").is_valid())\n",
```

Comment:
`tl.col("prompt")` -- I thought we settled on `tl["prompt]"` as the one way to do this?

Reply:
Yup